ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -44,66 +89,182 @@ descriptor :: PluginId -> PluginDescriptor IdeState
+    Just (name, ty) <- find (\(x, _) -> printName x == T.unpack identifier) localScope >>= \(name, mTy) -> (name,) <$> mTy,
+    Just TcModuleResult {tmrTypechecked = TcGblEnv {tcg_rdr_env, tcg_sigs}} <- mTmr,
+    -- not a top-level thing, to avoid duplication
+    not $ name `elemNameSet` tcg_sigs,
+    tyMsg <- showSDocForUser unsafeGlobalDynFlags (mkPrintUnqualified unsafeGlobalDynFlags tcg_rdr_env) $ pprSigmaType ty,
+    signature <- T.pack $ printName name <> " :: " <> tyMsg,
+    startCharacter <- _character _start,
+    startOfLine <- Position (_line _start) startCharacter,
+    beforeLine <- Range startOfLine startOfLine,
+    title <- if isQuickFix then "add signature: " <> signature else signature,
+    action <- TextEdit beforeLine $ signature <> "\n" <> T.replicate startCharacter " " =
+    [(title, [action])]
+  | otherwise = []
+
+sameThing :: SrcSpan -> Range -> Bool
+sameThing s1 s2 = (_start <$> srcSpanToRange s1) == (_start <$> Just s2)
+
+gblBindingTypeSigToEdit :: GlobalBindingTypeSig -> Maybe TextEdit
+gblBindingTypeSigToEdit GlobalBindingTypeSig {..}
+  | Just Range {..} <- srcSpanToRange $ getSrcSpan gbName,
+    startOfLine <- Position (_line _start) 0,
+    beforeLine <- Range startOfLine startOfLine =
+    Just $ TextEdit beforeLine $ T.pack gbRendered <> "\n"
+  | otherwise = Nothing
+
+data Mode
+  = -- | always displays type lenses of global bindings, no matter what GHC flags are set
+    Always
+  | -- | similar to 'Always', but only displays for exported global bindings
+    Exported
+  | -- |  follows error messages produced by GHC
+    Diagnostics
+  deriving (Eq, Ord, Show, Read, Enum)
+
+--------------------------------------------------------------------------------
+
+showDocRdrEnv :: DynFlags -> GlobalRdrEnv -> SDoc -> String
+showDocRdrEnv dflags rdrEnv = showSDocForUser dflags (mkPrintUnqualified dflags rdrEnv)
+
+data GetGlobalBindingTypeSigs = GetGlobalBindingTypeSigs
+  deriving (Generic, Show, Eq, Ord, Hashable, NFData, Binary)
+
+data GlobalBindingTypeSig = GlobalBindingTypeSig
+  { gbName     :: Name,
+    gbRendered :: String,
+    gbExported :: Bool
+  }
+
+gbSrcSpan :: GlobalBindingTypeSig -> SrcSpan
+gbSrcSpan GlobalBindingTypeSig {gbName} = getSrcSpan gbName
+
+newtype GlobalBindingTypeSigsResult = GlobalBindingTypeSigsResult [GlobalBindingTypeSig]
+
+instance Show GlobalBindingTypeSigsResult where
+  show _ = "<GetTypeResult>"
+
+instance NFData GlobalBindingTypeSigsResult where
+  rnf = rwhnf
+
+type instance RuleResult GetGlobalBindingTypeSigs = GlobalBindingTypeSigsResult
+
+rules :: Rules ()
+rules = do
+  define $ \GetGlobalBindingTypeSigs nfp -> do
+    tmr <- use TypeCheck nfp
+    -- we need session here for tidying types
+    hsc <- use GhcSession nfp
+    result <- liftIO $ gblBindingType (hscEnv <$> hsc) (tmrTypechecked <$> tmr)
+    pure ([], result)
+
+parseCustomConfig :: A.Object -> Maybe Mode
+parseCustomConfig = A.parseMaybe (\o -> read <$> o A..: "mode")
```

Comment:
This will crash if someone puts in a malformatted string, right? I think you might want some variant of safe read here --- perhaps `fmap fst . listToMaybe . reads`

Reply:
But a better play is probably just to write a `FromJSON` instance for `Mode` and delegate to that. The `Alternative` instance will let you parse more natural lower-case JSON strings.

Reply:
Thank you!

Reply:
Thank you!